### ISSUE TITLE
Output object types in R-Instat, how to process and display them at both R and .Net level

### ISSUE THREAD

**Issue by Patowhiz** [africanmathsinitiative/R-Instat#7778]
@dannyparsons @ChrisMarsh82 @lloyddewit @rdstern @lilyclements. 
Currently, we can only display 2 R outputs in the output viewer; **Graphs** and **Text**.

**Graph outputs**
These are displayed by internally running an extra R command, `grDevices::png`,  that takes a "snapshot" of the printed graph image with a specific resolution and saves it as a file that is then displayed in the output viewer.

**Text outputs**
These are displayed by internally running an extra R command, `utils::capture.output`, that takes the last R command and uses it as an input parameter to the `utils::capture.output` command. The results are then displayed in the output viewer.

In the .Net RLink, the detection of the expected output type is done by checking on; 

- Call type.
- Number of R script command lines. 
- Source of the R script. 

**My suggestions**

1. We detect the output type based on R-Instat "custom" R script commands, for instance we could have functions; `view_graph_object`, `view_text_object`, `view_html_object`. These R functions will detect if there is an R "viewer", if it's not there, like in the case of R-Instat, they save the object as a temporary file which can then be displayed in R-Instat. This will also simplify the dialogs implementations, developers won't need to change the `ICallType` for the different output types. All they need to do is use the custom R functions. 
2. At .Net level, we get all R outputs (the R scripts excluded) from the created files. This can be done by just getting the returned file path. This has the advantage of using less memory especially when the output viewer is implemented as a "_Recycle Viewer_". And also saving all the contents of the output viewer can be greatly simplified. This can also greatly improve the memory usage, we won't need to store big objects like html in the memory.

PR #7782 adds a html output type that can now be viewed in the output viewer. It also aims to demonstrate the above concepts.
Basically, dialogs that produce html objects will only need to call  `view_html_object` R function. That function checks of there is an R viewer, if not, creates a temporary file and returns it's file path. The at the .Net level, we file path is retrieved and the file is displayed in the output viewer.
This implementation can apply to graphs (by saving the graph objects with original resolution) and also text outputs.
I'm looking forward to your feedback.

**Comment by lloyddewit** [comment]
@Patowhiz Thanks for the chat this morning. I added some personal notes below. If I misunderstood anything, then please let me know.

@rdstern From the user perspective, the main potential advantage of this approach is that it would help users to run long intensive sessions, and then save the output window, without creating memory issues. The potential downside is that writing/reading temp files is much slower than writing/reading objects in memory. We would also need a reliable way of managing and cleaning up temp files. Both these potential downsides are manageable in my view.
What's your view on the relative priority of this PR? Is this something you'd like to discuss with @volloholic ?

Thanks

----
### Personal notes

> We detect the output type based on R-Instat "custom" R script commands, 

R-Instat would search the R script for specific strings. If for example, it found the name of the function that generates html output, then it would assume that the R script produces html output. In the future, we could do a safer check by using the RScript library to ensure that a specific function is actually called.

> for instance we could have functions; view_graph_object, view_text_object, view_html_object. These R functions will detect if there is an R "viewer", if it's not there, like in the case of R-Instat, they save the object as a temporary file which can then be displayed in R-Instat. 

R-Instat doesn't have an R viewer. However, the user may take script from the R-Instat log window and run it outside R-Instat in an environment that has an R viewer (e.g. RStudio). RStudio would then display the output correctly in its own viewer.

> This will also simplify the dialogs implementations, developers won't need to change the ICallType for the different output types. All they need to do is use the custom R functions.

We would no longer need the iCallType. Each dialog would know which type of output it expected and call the associated viewer function (e.g. `view_graph_object()`).

> At .Net level, we get all R outputs (the R scripts excluded) from the created files. This can be done by just getting the returned file path. 

Each output would be stored as a temp file. The output window would store a list of file names, rather than a list of objects.

> This has the advantage of using less memory especially when the output viewer is implemented as a "Recycle Viewer". 

The output window only stores a list of file names so even very long intensive sessions would not use much memory. A 'recycle viewer' would only store the currently visible output as objects in memory (plus some output just beyond the visible part to support smooth scrolling).

> And also saving all the contents of the output viewer can be greatly simplified. This can also greatly improve the memory usage, we won't need to store big objects like html in the memory.

We could save the contents window as a PDF, html etc. We could do this relatively easily, even for very long/intensive R-Instat sessions, because we are just constructing a single file from a list of txt/png/html files.

**Comment by Patowhiz** [comment]
@lloyddewit thanks for your feedback. You are correct.
I've also discussed this with @ChrisMarsh82. He generally agrees with the concept.

**Comment by dannyparsons** [comment]
@Patowhiz Thanks for this. I haven't been able to review this and all the implications carefully, but I can see some of the advantages to this approach.

However, I don't like having these functions `view_graph_object`, `view_text_object`, `view_html_object` now being external and part of the R script. We try to keep the R code as close to "normal" R code as possible and this is adding more custom R functions to every dialog. This is why `grDevices::png` and `utils::capture.output` are called internally.

I'm happy with your new functions to capture these outputs for the new viewer, but can this still be done internally so that the R scripts remain the same?

**Comment by Patowhiz** [comment]
@dannyparsons that principle is well noted. 
The downside to that approach is how do we detect and respond accordingly when the script is run in the script window? I was leaning towards a solution where, an output script run by the dialog behaves in a similar manner even when run in the script window. 
Before, `grDevices::png` was not being called when the script was being run in the output window, that's why the RViewer popped up, which meant that output was no longer available in the output window. Also, with the new implementation of getting our outputs as files that will be an issue. 
 
I also managed to merge the 3 functions into just one, `view_object` . Which is all done on behalf of the dialogs.

I'll explore further to see if we can get a work around.

**Comment by Patowhiz** [comment]
In addition as indicated in my code comments. R-Studio has a way of detecting when the output should go to their output viewer. 
I'm still exploring how that can be possible in R-Instat. One think that is clear when investigating this, for every image or graph output, R-Studio creates a temporary file. When looking into the file, it is what's displayed in their viewer.

**Comment by lloyddewit** [comment]
@dannyparsons @Patowhiz Thank you for the comments above.

I understand that we want the R code in the output window to be as generic as possible so that it could potentially be executed outside R-Instat (e.g. in RStudio). Therefore, we want to minimise calling R-Instat specific functions (such as functions that determine the output destination for graphs, tables, etc.).

Would the following behaviour be acceptable?
- The script window executes the visible script and, if not otherwise specified in the visible script, sends the output to the default destination.
- The R-Instat dialog has more options regarding the output destination, but these are specified in internal scripts that do not appear in the output window.

I'm not an R expert and it's possible I misunderstood your comments above, if so, then please let me know.
Thanks

**Comment by Patowhiz** [comment]
@lloyddewit yes you understood the issue correctly.

> The script window executes the visible script and, if not otherwise specified in the visible script, sends the output to the default destination

This is precisely what I'm trying to avoid. For instance, as a user when I execute a script that produces a html output, I expect it to be displayed within R-Instat, regardless of the source of the execution. To me that consistency is important. Reason being, as a user I may want to share my outputs, and it's now trivially easy to do that in R-Instat. If the outputs go to another destination, like the browser, the user will need to learn about that destination too. I think that's a poor user experience. Remember we still want to support other output types too, for instance latex, pdfs, that means more default destinations. In addition, the user may be lacking the destination software for viewing the output.

Determining what the output will be without a specific function check proved to be difficult in my research. RStudio seems to have figured how to do this, but again almost all packages are expected to be run inside RStudio!.

@rdstern @lloyddewit @dannyparsons for now, It's either;
1. We work with this 1 function as we keep thinking around how to avoid it. This adds 1 extra foreign function to the user (a demerit) but guarantees consistency in user flow and expectation. Also simplifies our R link implementation (merits).
2. Or we remove it and accept the demerits. Lack of consistency in user flow and expectation. Also users should have software that support the output should the destination default to what R chooses.  

I suggest option 1.
Thanks.

**Comment by dannyparsons** [comment]
I would also like to have consistency with code run through the script window, however, unfortunately, I don't think the benefits of this outweigh modify the script that every dialog produces. Although the script window is important, it's still an "advanced" feature, and not used by all.

So for now I would go for option 2, and be looking for ways of having the detection done through the script window e.g. crude ways by guessing the output type based on string matching ("ggplot" - graph), or user specified in the script window, or something else.

I think if it proves too difficult to replicate what RStudio has managed to do in this regard, then maybe we have to accept another solution, but I would prefer not to go straight to that.

I'm happy for @volloholic to overrule me if he feels I'm being too insistent on the R script being unmodified.

**Comment by dannyparsons** [comment]
Also, the proposed method 1 will only work if the script in the script window is generated from a dialog and includes the view function at the end. If the user writes their own script we will still need some detection of the output, as plain R won't have these view functions.

**Comment by rdstern** [comment]
I have had a very constructive discussion with @Patowhiz and try to give my understanding of the differences in the 2 approaches.

Here is a boxplot for the survey data with the current (David/Danny) script and the new (Patrick) script.  

Danny/David:
```
# Code généré par la boîte de dialogue, Boxplot

survey <- data_book$get_data_frame(data_name="survey")
box_plot1 <- ggplot2::ggplot(data=survey, mapping=ggplot2::aes(y=yield, x=variety, fill=fertgrp)) + ggplot2::geom_boxplot(varwidth=TRUE, outlier.colour="red") + theme_grey()
data_book$add_graph(graph_name="box_plot1", graph=box_plot1, data_name="survey")
data_book$get_graphs(data_name="survey", graph_name="box_plot1")
rm(list=c("box_plot1", "survey"))

```
And the Patrick code:
```
# Code généré par la boîte de dialogue, Boxplot Options

survey <- data_book$get_data_frame(data_name="survey")
box_plot <- ggplot2::ggplot(data=survey, mapping=ggplot2::aes(y=yield, x=variety, fill=fertgrp)) + ggplot2::geom_boxplot(outlier.colour="red") + theme_grey()
data_book$add_object(data_name="survey", object_name="box_plot", object_type_label="graph", object_format="image", object=box_plot, internal=FALSE)
view_object(object=box_plot, object_format="image")
rm(list=c("box_plot", "survey"))

```
So my (limited) understanding is as follows:
a) Lines 1 and 2 above are essentially unchanged.
b) Line 3 adds the graph (David/Danny) or the object into the data book.  There seems to be general happiness about the way the graph addition is generalised into an object addition.  I am also very happy that we are getting back to a systematic treatment of the objects produced.

I have always been keen on having improved facilities for handling the objects, and so improving on the Prepare > R Objects > View, Rename, Reorder, Delete.  Currently they are inconsistent, e.g. View and Reorder don't include all the types of object, while they are there in Reorder and Delete.  I have been at a slight loss to know why Danny/David never seemed keen on improvements here, so find it great that this is now been taken up.  (Maybe I am just misunderstanding David/Danny's priorities here?)

I hope these improvements will extend to the storage locations of these objects, because currently the graphs all disappear after the current session.  This seems unprofessional - and caused problems to the AIMS students on the recent climatic course.  I am at a slight loss as to why resolution of this issue is not being addressed urgently.  

So, my understanding is that there is no conceptual issue on the changes in line 3.
c) Line 4 is either (David/Danny):
`data_book$get_graphs(data_name="survey", graph_name="box_plot1")`
or (Patrick)
`view_object(object=box_plot, object_format="image")`
I don't quite understand why there seems to be a big deal here?  As I understand it, the David/Danny function is very clear on what you are getting, i.e. box_plot1, from the survey data frame, but hides what you then are going to do with it.  I gather there is another argument and the default is to show it in the R viewer.  I assume that will need to change, because the R-viewer is giving us problems and one advantage of the new output window is that the other options (html or plotly) become more obvious.  
In contrast, the Patrick function hides some parts of what you get - and you don't need to go to the data book to get it, because you only just put it there.  It is much clearer, even from the name of what you are doing with it.
I would like the functions to be usable as generally as possible, so would like the view _object to be able to get an object from the data book, without assuming it is the one that was just generated - but I am happy if the default is to omit those argument, if it was just put there.
I also don't mind if this changes to a get_object with extra arguments to say what you do once you have got it.  But if that makes the function complicated, then a few functions, such as view_object, delete_object, might be fine too.

I assume I must have missed something important here?  Maybe @lloyddewit understands the problems better?

**Comment by lloyddewit** [comment]
> Maybe @lloyddewit understands the problems better?

likely not!
The only thing I would add is that `view_object()` is R-Instat specific and won't work in other environments. Also, as @dannyparsons pointed out, the script window should be able to send output to a sensible location just using standard R.
I personally like the idea of the output window just showing the 'real' R script (not the R-Instat specific internal code) but I don't have strong views. I'm happy with whatever you decide.

**Comment by Patowhiz** [comment]
Okay.
I'll remove the `view_object` and replace it with `get_object`, just like before.
So the sequence of the script will be the same.
`get_object` will call view_object internally, the user will not know that.

**Comment by rdstern** [comment]
Aaah, has @lloyddewit finally told me what the situation is.  If I now understand then the get_object function would also work in RStudio.  And `view` might be currently an invisible argument.  If so then, in general there may be various options in `view` that would only work in R-Instat, but at least some (maybe only the default, or different ones) would also work in rstudio.  That would be fine.

**Comment by Patowhiz** [comment]
No. The view will always work in any R environment.  It's designed to do so.. 
I also don't understand what real R script means.
I'm agreeing to revert back to get_object even though I strongly think it makes the script inefficient, because I'm assuming performance will always.be negligible..
